src/agreements/CollateralAgreement.sol

```diff
@@ -98,22 +97,25 @@ contract CollateralAgreementFramework is IAgreementFramework {
         emit AgreementCreated(agreementId, params.termsHash, params.criteria);
     }
 
-    function joinAgreement(uint256 id, uint256 balance) external {
-        if (balance < agreement[id].criteria) revert InsufficientBalance();
-        SafeTransferLib.safeTransferFrom(token, msg.sender, address(this), balance);
+    function joinAgreement(
+        uint256 id,
+        CriteriaResolver calldata resolver
+    ) external override {
+        if (_isPartOfAgreement(id, msg.sender)) revert PartyAlreadyJoined();
+        if (msg.sender != resolver.party) revert PartyMustMatchCriteria();
 
-        uint256 newPartyId = agreement[id].party.length;
+        _validateCriteria(agreement[id].criteria, resolver);
 
-        Position memory newPositon = Position(newPartyId, balance, PositionStatus.Idle);
+        SafeTransferLib.safeTransferFrom(token, msg.sender, address(this), resolver.balance);
 
-        agreement[id].party.push(msg.sender);
-        agreement[id].position[msg.sender] = newPositon;
+        _addPosition(id, PositionParams(msg.sender, resolver.balance));
 
-        emit AgreementJoined(id, msg.sender, balance);
+        emit AgreementJoined(id, msg.sender, resolver.balance);
     }
 
-    function finalizeAgreement(uint256 id) external {
-        if (agreement[id].party[agreement[id].position[msg.sender].id] != msg.sender)
+    function finalizeAgreement(uint256 id) external override {
```

Comment:
Maybe add some NatSpec comments to both `finalizeAgreement` and `joinAgreement` so that users/developers will understand the difference between the two?

Reply:
Just realized that there are already comments in the `IAgreementFramework` interface, so ignore my previous comment 😆